onnxruntime/core/providers/cpu/tensor/transpose.cc

```diff
@@ -52,35 +94,38 @@ static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target
                             size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
                             const uint8_t* source, uint8_t* target, size_t element_size) {
   size_t blocksize = num_elts_in_block * element_size;
+  MultiIndex* mindex = (MultiIndex*)alloca(num_axes * sizeof(MultiIndex));
+  size_t naxes = IncrementIndexAndComputeOffsetSetup(mindex, num_axes, target_dims, stride, element_size);
+
   // index used to iterate over target iteration-space
   std::vector<int64_t> target_index(num_axes, 0);
+  const uint8_t* local_source = source;
   for (size_t i = 0; i < num_blocks; ++i) {
-    // convert target_index into an offset in source data
-    size_t source_offset = ComputeOffset(target_index, stride, num_axes);
-
     // copy
-    memcpy(target, source + source_offset * element_size, blocksize);
+    memcpy(target, local_source, blocksize);
 
     // increment target_index:
-    IncrementIndex(target_index, target_dims, num_axes);
+    IncrementIndexAndComputeOffset(mindex, naxes, local_source);
     target += blocksize;
   }
 }
 
 static void DoTransposeImpl(int64_t num_axes, const std::vector<int64_t>& target_dims,
                             size_t num_blocks, size_t num_elts_in_block, const std::vector<size_t>& stride,
                             const std::string* source, std::string* target) {
+  ORT_ENFORCE(num_axes > 0, "Transpose not implemented for empty tensors.");
```

Comment:
For 1-D tensors, can we just make a copy of the input tensor's contents and return ?

Reply:
If there's only one axis won't IsReshape return true so this function won't be called? If so, we don't need the ORT_ENFORCE.

---
In reply to: [520125596](https://github.com/microsoft/onnxruntime/pull/5550#discussion_r520125596) [](ancestors = 520125596)

Reply:
True